### PR TITLE
Ensure no blank `Sec-Websocket-Protocol` headers and warn if websocket subprotocol edge case occur

### LINKED ISSUE #332: `./contrib` docs, theia: Theia proxy hangs, where Thiea on own port works correctly
## Summary

I'm trying to install and run Theia via [these instructions](https://jupyter-server-proxy.readthedocs.io/en/latest/convenience/packages/theia.html).

I have what I believe to be a minimal installation of JupyterHub / Theia working, via [this Dockerfile](https://github.com/nipraxis/nipraxis-images/blob/main/minimal-theia-proxy/Dockerfile).  The image is at <https://hub.docker.com/r/matthewbrett/minimal-theia-proxy>. 

I can get the Theia interface when running Theia from the terminal, but not when executed via `jupyter_theia_proxy`.

## Detail

I can show that Theia is working correctly outside the JupyterHub Theia proxy by doing the following:

```
docker run -ti --rm -p 8888:8888 -p 3000:3000 matthewbrett/minimal-theia-proxy:82534d5
```

This gives the JupyterLab interface available on localhost with the link advertised in the terminal.   Using the JupyterLab terminal launcher, I run the following command (see below for the `theia` running script):

```
theia start . --hostname 0.0.0.0 --port=3000
```

At that point, I can get a functional Theia interface at http://localhost:3000 . 

However, if instead, I click on the Theia launcher, I see that the browser is trying to load http://127.0.0.1:8888/theia/ - but this hangs at what appears to be the Theia loading stage, with a dark background and spinning gray semicircle - the same thing that very briefly preceeds the successful loading of Theia above.  However, in this case, the browser hangs there, and never reaches the Theia interface. 

Meanwhile, in the console from which I ran the Docker run command, I get repeated messages of form:

```
[I 2022-03-09 17:24:11.223 ServerApp] 302 GET /socket.io/?EIO=4&transport=polling&t=NzljQir (172.17.0.1) 0.54ms
[W 2022-03-09 17:24:11.233 ServerApp] 404 GET /socket.io?EIO=4&transport=polling&t=NzljQir (172.17.0.1) 4.17ms referer=http://127.0.0.1:8888/theia/
```

Can you point me to where I am going wrong in using the Theia proxy here?   How can I further debug?

## Theia loading command

I found that the instructions at <https://jupyter-server-proxy.readthedocs.io/en/latest/convenience/packages/theia.html> could not work for my installation, in that the command `<my-theia-build-path>/node_modules/.bin/theia` will not start Theia from an arbitrary directory.  I had to wrap the `theia` command in a shell wrapper, in order to run the command from a location where it could find the `package.json` file:

<https://github.com/nipraxis/nipraxis-images/blob/e2dcf0ad749c7bee80a99f1fd4cbbdaf548090de/nipraxis-hub/theia-bin>

## Theia hostname

The command being run by `jupyter_theia_proxy` appears to be:

```python
['theia', 'start', '.', '--hostname 127.0.0.1', '--port=' + str(port)]
```

This differs from my command run from the terminal, in that the terminal command uses `--hostname 0.0.0.0`. 

I've tried replacing `--hostname 0.0.0.0` into the `jupyter_theia_proxy` command, with no benefit.

### LINKED ISSUE #442: Handler subprotocol method fails to detect empty list
### Bug description

Hello JSP developers!

My colleague @zoghbi-a and I are attempting to proxy socket.io using JSP and have run into a subprotocol issue that I've started to track down. Please note I am coming as a JSP user with not much networking experience. I'd like to open a discussion about this issue and would be happy to be told I'm just not using JSP properly and corrected. I also didn't want to open a PR in the event I'm misunderstanding something. Thank you for your patience!

Effectively, we're receiving strange headers on our socket.io server that are resulting in a HTTP 400 Bad Request. Tracking down the HTTP response being sent by the proxy, I noticed the following fields being sent in the HTTP header:

<details>

```
> e:\heasarc\gitrepos\scieserver-js9\envjpyjs9\lib\site-packages\jupyter_server_proxy\handlers.py(503)start_websocket_connection()
(Pdb) print(request.headers)
Host: localhost:8888
Connection: Upgrade
Pragma: no-cache
Cache-Control: no-cache
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36
Accept-Language: en-US,en
Upgrade: websocket
Origin: http://localhost:8888
Sec-Websocket-Version: 13
Accept-Encoding: gzip
Cookie: username-localhost-8888="<irrelevant values>"
Sec-Websocket-Key: b'XXXXXXXXXXXXXXXXXXXX'
Sec-Websocket-Extensions: permessage-deflate; client_max_window_bits
Sec-Websocket-Protocol:
```

</details>

Specifically, I noticed that last line:

```
Sec-Websocket-Protocol:
```

As I understand it, when a websocket protocol is not specified, the `Sec-Websocket-Protocol` is supposed to be omitted, rather than sent without a value. 

The following is my best attempt to understand how this condition is arising (all links are permalinked to the current `main` commit: `a65437a`):

The HTTP 400 error is originating from the following HTTPRequest in JSP in attempting to start a websocket connection: 
https://github.com/jupyterhub/jupyter-server-proxy/blob/a65437a54c4069a6de20c9845b41794cececdd75/jupyter_server_proxy/handlers.py#L492-L497

The cause is the value of `self.subprotocols`. It's receiving an empty list `[]`. This empty list is originating from:
https://github.com/tornadoweb/tornado/blob/65a9e48f8ce645f104e3e0aa772222e70b0376d9/tornado/websocket.py#L902
most likely due to a subprotocol header not being provided on our end. This harkens back to another user's issue: https://github.com/jupyterhub/jupyter-server-proxy/issues/47

Following the trace, the JSP handler's `select_subprotocol` appears to be processing the empty list properly:
https://github.com/jupyterhub/jupyter-server-proxy/blob/a65437a54c4069a6de20c9845b41794cececdd75/jupyter_server_proxy/handlers.py#L533-L539

by properly falling back on Tornado's WebSocketHandler, which properly returns `None`:
https://github.com/tornadoweb/tornado/blob/65a9e48f8ce645f104e3e0aa772222e70b0376d9/tornado/websocket.py#L337-L360

However, `select_subprotocol` also caches the provided value:
https://github.com/jupyterhub/jupyter-server-proxy/blob/a65437a54c4069a6de20c9845b41794cececdd75/jupyter_server_proxy/handlers.py#L535

which is what is used websocket connection method:
https://github.com/jupyterhub/jupyter-server-proxy/blob/a65437a54c4069a6de20c9845b41794cececdd75/jupyter_server_proxy/handlers.py#L496
thereby ignoring the "sanitized" subprotocol value provided by Tornado's WebSocketHandler.

## TLDR solution:
I feel that either the `jupyter_server_proxy.handlers.ProxyHandler.proxy_open` should be directly using the sanitized result of `jupyter_server_proxy.handlers.ProxyHandler.select_subprotocol`:
```
self.ws = await pingable_ws_connect(
    request=request,
    on_message_callback=message_cb,
    on_ping_callback=ping_cb,
    subprotocols=self.select_subprotcol(self.subprotocols),
    resolver=resolver,
)
```

or, `jupyter_server_proxy.handlers.ProxyHandler.select_subprotocol` should instead use the Tornado WebSocketHandler's cached value of the final selected subprotocol: https://github.com/tornadoweb/tornado/blob/65a9e48f8ce645f104e3e0aa772222e70b0376d9/tornado/websocket.py#L903

```
self.ws = await pingable_ws_connect(
    request=request,
    on_message_callback=message_cb,
    on_ping_callback=ping_cb,
    subprotocols=self.selected_subprotcol,
    resolver=resolver,
)
```

Please let me know your thoughts! I appreciate your time and effort on this project :)

### LINKED ISSUE #445: PDF preview of LaTeX Workshop (code-server) results in `HTTP 400: Bad Request`
### Bug description

code-server: https://github.com/coder/code-server/issues/6667

JupyterLab:

```
[I 2024-02-13 15:50:09.573 ServerApp] Trying to establish websocket connection to ws://localhost:56227/proxy/39607/
[E 2024-02-13 15:50:09.579 ServerApp] Uncaught exception GET /code-server/proxy/39607/ (192.168.65.1)
    HTTPServerRequest(protocol='http', host='127.0.0.1:8888', method='GET', uri='/code-server/proxy/39607/', version='HTTP/1.1', remote_ip='192.168.65.1')
    Traceback (most recent call last):
      File "/usr/local/lib/python3.11/site-packages/tornado/websocket.py", line 940, in _accept_connection
        await open_result
      File "/usr/local/lib/python3.11/site-packages/jupyter_server_proxy/handlers.py", line 853, in open
        return await super().open(path)
               ^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/local/lib/python3.11/site-packages/jupyter_server_proxy/handlers.py", line 678, in open
        return await super().open(self.port, path)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/local/lib/python3.11/site-packages/jupyter_server_proxy/handlers.py", line 557, in open
        return await self.proxy_open("localhost", port, proxied_path)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/local/lib/python3.11/site-packages/jupyter_server_proxy/handlers.py", line 505, in proxy_open
        await start_websocket_connection()
      File "/usr/local/lib/python3.11/site-packages/jupyter_server_proxy/handlers.py", line 492, in start_websocket_connection
        self.ws = await pingable_ws_connect(
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^
    tornado.httpclient.HTTPClientError: HTTP 400: Bad Request
```

### How to reproduce

1. [Terminal] Command `docker run -it --rm -p 8888:8888 -p 47395:47395 glcr.b-data.ch/jupyterlab/r/verse:latest-devtools`
2. [Browser] Open http://127.0.0.1:8888/lab?token=[redacted]
3. [JupyterLab] Click on 'code-server [↗]'
4. [code-server] When asked 'Do you trust the authors of the files in this folder?'
    - Click on 'Yes, I trust the authors'
5. [code-server] Explorer > Right click > New File...: `sample.tex`
    - Use sample from **Code sample**
7. [code-server] View > Command Palette... (F1): LaTeX Workshop: Build LaTeX project
8. [code-server] View > Command Palette... (F1): LaTeX Workshop: View LaTeX PDF file in VSCode tab
9. [code-server] File `sample.tex`: Change `\date{2024-02-09}` to `\date{2024-02-13}`

#### Expected behaviour

The LaTeX PDF file in VSCode tab updates accordingly.

#### Actual behaviour

The LaTeX PDF file in VSCode tab does not update.

### Code sample

Sample $\TeX$ file:

```tex
\documentclass{article}

\title{My first document}
\date{2024-02-09}
\author{John Doe}

\begin{document}
  \maketitle
  \newpage

  Hello World!

\end{document}
```

### Your personal set up

 - OS:
    - Host: macOS 14.3.1 (Sonoma)
    - Container: Debian 12 (bullseye)
 - Version(s):
    - JupyterHub: 4.0.2
    - JupyterLab: 4.1.0
    - Jupyter Server Proxy: 4.1.0
    - code-server: 4.21.1 (Code 1.86.1)
    - LaTeX Workshop: 9.18.0

<details><summary>Full environment</summary>

Container:

```
aiohttp==3.9.3
aiosignal==1.3.1
alembic==1.13.1
anyio==4.2.0
argon2-cffi==23.1.0
argon2-cffi-bindings==21.2.0
arrow==1.3.0
astroid==3.0.3
asttokens==2.4.1
async-generator==1.10
async-lru==2.0.4
attrs==23.2.0
autopep8==2.0.4
Babel==2.14.0
beautifulsoup4==4.12.3
bleach==6.1.0
certifi==2024.2.2
certipy==0.1.3
cffi==1.16.0
charset-normalizer==3.3.2
colorama==0.4.6
comm==0.2.1
cryptography==42.0.2
debugpy==1.8.1
decorator==5.1.1
defusedxml==0.7.1
dill==0.3.8
docstring-to-markdown==0.13
executing==2.0.1
fastjsonschema==2.19.1
flake8==7.0.0
fqdn==1.5.1
frozenlist==1.4.1
gitdb==4.0.11
GitPython==3.1.41
greenlet==3.0.3
h11==0.14.0
httpcore==1.0.2
httpx==0.26.0
idna==3.6
importlib-metadata==7.0.1
ipykernel==6.29.2
ipython==8.21.0
ipython-genutils==0.2.0
isoduration==20.11.0
isort==5.13.2
jedi==0.19.1
Jinja2==3.1.3
json5==0.9.14
jsonpointer==2.4
jsonschema==4.21.1
jsonschema-specifications==2023.12.1
jupyter-events==0.9.0
jupyter-lsp==2.2.2
jupyter-server-mathjax==0.2.6
jupyter-telemetry==0.1.0
jupyter_client==8.6.0
jupyter_core==5.7.1
jupyter_server==2.12.5
jupyter_server_proxy==4.1.0
jupyter_server_terminals==0.5.2
jupyterhub==4.0.2
jupyterlab==4.1.0
jupyterlab-lsp==5.0.3
jupyterlab_git==0.50.0
jupyterlab_pygments==0.3.0
jupyterlab_server==2.25.2
Mako==1.3.2
MarkupSafe==2.1.5
matplotlib-inline==0.1.6
mccabe==0.7.0
mistune==3.0.2
multidict==6.0.5
nbclassic==1.0.0
nbclient==0.9.0
nbconvert==7.16.0
nbdime==4.0.1
nbformat==5.9.2
nest-asyncio==1.6.0
notebook==7.0.7
notebook_shim==0.2.3
oauthlib==3.2.2
overrides==7.7.0
packaging==23.2
pamela==1.1.0
pandocfilters==1.5.1
parso==0.8.3
pexpect==4.9.0
platformdirs==4.2.0
pluggy==1.4.0
prometheus-client==0.19.0
prompt-toolkit==3.0.43
psutil==5.9.8
ptyprocess==0.7.0
pure-eval==0.2.2
pycodestyle==2.11.1
pycparser==2.21
pydocstyle==6.3.0
pyflakes==3.2.0
Pygments==2.17.2
pylint==3.0.3
pyOpenSSL==24.0.0
python-dateutil==2.8.2
python-json-logger==2.0.7
python-lsp-jsonrpc==1.1.2
python-lsp-server==1.10.0
pytoolconfig==1.3.1
PyYAML==6.0.1
pyzmq==25.1.2
radian==0.6.12
rchitect==0.4.6
referencing==0.33.0
requests==2.31.0
rfc3339-validator==0.1.4
rfc3986-validator==0.1.1
rope==1.12.0
rpds-py==0.17.1
ruamel.yaml==0.18.6
ruamel.yaml.clib==0.2.8
Send2Trash==1.8.2
simpervisor==1.0.0
six==1.16.0
smmap==5.0.1
sniffio==1.3.0
snowballstemmer==2.2.0
soupsieve==2.5
SQLAlchemy==2.0.26
stack-data==0.6.3
terminado==0.18.0
tinycss2==1.2.1
tomli==2.0.1
tomlkit==0.12.3
tornado==6.4
traitlets==5.14.1
types-python-dateutil==2.8.19.20240106
typing_extensions==4.9.0
ujson==5.9.0
uri-template==1.3.0
urllib3==2.2.0
wcwidth==0.2.13
webcolors==1.13
webencodings==0.5.1
websocket-client==1.7.0
whatthepatch==1.0.5
yapf==0.40.2
yarl==1.9.4
zipp==3.17.0
```
</details>

</details>

<details><summary>Logs</summary>

See **Bug description**.

</details>

### PR DESCRIPTION
**Pull request by consideRatio**
This PR ensures we don't pass a blank `Sec-Websocket-Protocol` header ([Sending a blank header is incorrect](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_servers#subprotocols)) when we proxy a websocket. This fixes #442, fixes #445, is assumed to fix #332, and closes #448.

This project currently proxies websockets by first accepting an incomming websocket connection were a subprotocol is established, and then opening a websocket connection to a server to proxy to. This is a problem and can lead to bugs that we now warn about. This PR adds tests to document our current problematic handling.

Thank you @duytnguyendtn, @zoghbi-a, and @benz0li for working this - this is building a lot on their investigation and conclusions, which were critical as I didn't understand much about this before.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -50,6 +50,8 @@ dependencies = [
     "importlib_metadata >=4.8.3 ; python_version<\"3.10\"",
     "jupyter-server >=1.0",
     "simpervisor >=1.0",
+    "tornado >=5.1",
```
Comment (duytnguyendtn) [inline pyproject.toml:53]: Does removing the tornado 4.5.x compatibility patch and setting the min version of tornado >=5.1 make this a breaking change? Trying to foresee which version of JSP this patch will be released with

Comment (consideRatio) [reply pyproject.toml:53]: No, this is from #457, the PR is based on it for the moment.

Comment (consideRatio) [reply pyproject.toml:53]: As long as we don't force a new major version in the dependency, i figure it should be fine at least.

**Comment by benz0li** [comment]
@consideRatio I merged `main` and `consideRatio/test-subprotocols` into `benz0li/fix-442-and-445-simple`, adapted the Proxy tests and ran the test suite: It also passes the *empty* and *no* subrotocol tests.

**Review by minrk — APPROVED** [review]
